staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -1219,6 +1222,223 @@ func RunSendInitialEventsBackwardCompatibility(ctx context.Context, t *testing.T
 	w.Stop()
 }
 
+func RunWatchSemantics(ctx context.Context, t *testing.T, store storage.Interface) {
+	trueVal, falseVal := true, false
+	addEventsFromCreatedPods := func(createdInitialPods []*example.Pod) []watch.Event {
+		var ret []watch.Event
+		for _, createdPod := range createdInitialPods {
+			ret = append(ret, watch.Event{Type: watch.Added, Object: createdPod})
+		}
+		return ret
+	}
+	initialEventsEndFromLastCreatedPod := func(createdInitialPods []*example.Pod) watch.Event {
+		return watch.Event{
+			Type: watch.Bookmark,
+			Object: &example.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: createdInitialPods[len(createdInitialPods)-1].ResourceVersion,
+					Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+				},
+			},
+		}
+	}
+	scenarios := []struct {
+		name                string
+		allowWatchBookmarks bool
+		sendInitialEvents   *bool
+		resourceVersion     string
+
+		initialPods                func(ns string) []*example.Pod
+		podsAfterEstablishingWatch func(ns string) []*example.Pod
```

Comment:
no, please let's leave it as it is.
It is easier to follow because it is immediately clear that the ns is provided instead of reading the test body/setup

Reply:
I still think what I'm proposing is cleaner - I don't think that you get this info from the test.

But for the sake of making progress (and given it's working), let's merge it - I will open a cleanup PR where we can discuss it further.

Reply:
okay, np, i can open a follow-up pr.